src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs

```diff
@@ -322,6 +324,299 @@ public virtual void Encode(TextWriter output, char[] value, int startIndex, int
             }
         }
 
+        public unsafe virtual OperationStatus Encode(ReadOnlySpan<char> source, Span<char> destination, out int charsConsumed, out int charsWritten, bool isFinalBlock = true)
+        {
+            // Optimization: Detect how much "doesn't require escaping" data exists at the beginning of the buffer,
+            // and memcpy it directly to the destination.
+
+            int idx = FindFirstCharacterToEncode(source);
+
+            int numCharsToCopy = idx;
+            if (numCharsToCopy < 0)
+            {
+                numCharsToCopy = source.Length;
+            }
+
+            if (!source.Slice(0, numCharsToCopy).TryCopyTo(destination))
```

Comment:
>Slice [](start = 24, length = 5)

do we need to check if numCharsToCopy == 0 before we do this operation?

Reply:
No, the call to `Slice` will detect it and no-op. There's also a unit test for this specific case: `EncodeUtf8_EmptyInput_AlwaysSucceeds`.